### PR TITLE
Implement runtime suspension for OSX

### PR DESCRIPTION
**Pull request by janvorli**
This change implements runtime suspension for OSX using activation
injection, the same mechanism as Linux uses.
I have modified the activation injection mechanism by adding a new
function that CoreCLR registers with PAL and PAL calls it to check
if an instruction address is safe for injection. The injection is
performed only if this new function returns true.
The activation on OSX is performed by suspending the target thread,
checking the instruction address in the suspended thread's context
using the above mentioned function and modifying the context to
perform the injection only at injection safe place.

As an additional detail, I have fixed the #1594, since it was similar to the activation handler wrapper creation.

### CODE REVIEW

**Comment by janvorli** [comment]
@jkotas, @kangaroo Can you take a look please?

File: src/pal/src/include/pal/thread.hpp [hunk 1]
```diff
@@ -314,6 +314,10 @@ namespace CorUnix
         DWORD m_dwLwpId;
         pthread_t m_pthreadSelf;        
 
+#if HAVE_MACH_THREADS
+        mach_port_t m_machPortSelf;
```
Comment (kangaroo) [inline src/pal/src/include/pal/thread.hpp]: I don't see this being cleaned up when the thread dies, which means we'll leak the port.

**Comment by janvorli** [comment]
cc: @sergiy-k

File: src/pal/src/include/pal/thread.hpp (hunk 1 above)
Comment (janvorli) [inline src/pal/src/include/pal/thread.hpp]: Are you sure it needs a cleanup? I was under an impression that the pthread_mach_thread_np just exposes mach port that is stored in the pthread_t and is bound to it.

Comment (kangaroo) [inline src/pal/src/include/pal/thread.hpp]: You're correct, I just checked and it isn't cloned or ref'd so this should be ok.

**Comment by sergiy-k** [comment]
LGTM

**Comment by jkotas** [comment]
LGTM

**Comment by desunit** [comment]
Thanks so much! Compiled successfully but I had to install libicu-dev first: `brew install icu4c`

**Pull request merged**